First we'll read the data for each of the 4 datasets

In [1]:
import pandas as pd
import os

project_root = os.getcwd()
kiva_loans_path = os.path.join(project_root, "../../kiva_data/kiva_loans.csv")
kiva_loans = pd.read_csv(kiva_loans_path)

kiva_loans.head()

id  funded_amount  loan_amount             activity          sector  \
0  653051          300.0        300.0  Fruits & Vegetables            Food   
1  653053          575.0        575.0             Rickshaw  Transportation   
2  653068          150.0        150.0       Transportation  Transportation   
3  653063          200.0        200.0           Embroidery            Arts   
4  653084          400.0        400.0           Milk Sales            Food   

                                                 use country_code   country  \
0            To buy seasonal, fresh fruits to sell.            PK  Pakistan   
1  to repair and maintain the auto rickshaw used ...           PK  Pakistan   
2  To repair their old cycle-van and buy another ...           IN     India   
3  to purchase an embroidery machine and a variet...           PK  Pakistan   
4                           to purchase one buffalo.           PK  Pakistan   

         region currency  partner_id                posted_time  \
0        Lahore      PKR       247.0  2014-01-01 06:12:39+00:00   
1        Lahore      PKR       247.0  2014-01-01 06:51:08+00:00   
2     Maynaguri      INR       334.0  2014-01-01 09:58:07+00:00   
3        Lahore      PKR       247.0  2014-01-01 08:03:11+00:00   
4  Abdul Hakeem      PKR       245.0  2014-01-01 11:53:19+00:00   

              disbursed_time                funded_time  term_in_months  \
0  2013-12-17 08:00:00+00:00  2014-01-02 10:06:32+00:00            12.0   
1  2013-12-17 08:00:00+00:00  2014-01-02 09:17:23+00:00            11.0   
2  2013-12-17 08:00:00+00:00  2014-01-01 16:01:36+00:00            43.0   
3  2013-12-24 08:00:00+00:00  2014-01-01 13:00:00+00:00            11.0   
4  2013-12-17 08:00:00+00:00  2014-01-01 19:18:51+00:00            14.0   

   lender_count                          tags borrower_genders  \
0            12                           NaN           female   
1            14                           NaN   female, female   
2             6  user_favorite, user_favorite           female   
3             8                           NaN           female   
4            16                           NaN           female   

  repayment_interval        date  
0          irregular  2014-01-01  
1          irregular  2014-01-01  
2             bullet  2014-01-01  
3          irregular  2014-01-01  
4            monthly  2014-01-01

In [2]:
mpi_url = "https://storage.googleapis.com/kagglesdsdata/datasets/12414/19728/kiva_mpi_region_locations.csv?X-Goog-Algorithm=GOOG4-RSA-SHA256&X-Goog-Credential=gcp-kaggle-com%40kaggle-161607.iam.gserviceaccount.com%2F20250604%2Fauto%2Fstorage%2Fgoog4_request&X-Goog-Date=20250604T200827Z&X-Goog-Expires=259200&X-Goog-SignedHeaders=host&X-Goog-Signature=85afbaf15a1cb028b110d83362747be45da1e52966ccad4736d81967bf10b482985f2d9abf9df8c4561789bf43684741de3f2cd077256c5275e0a4079042dda8c82457f8f6e3b85db203c7acb56bde0c5256cc4ce573d5b547661c0e52310d9fd0a8b0eca074098c275fa5e82ca3207c01b15cf88e1b897f6520ea9b82d0af90e7521afc2016ab3b8e8c61eff98f725c6c721a5b8ce2aa18a58e8ef01313c912b6a3f040df81e86352d6e26427ef1f55b4dee41bcdb3059e22a3f5e7f555c15cdf8eb1fb9ceffb307c05a075aa4acd1119192d1bfe86d684831658f77e897acdc230224dea2cf9c0f5eda1349b2356a7e6b4b5296a262b3aa5c00d2744826323"

mpi_df = pd.read_csv(mpi_url)

mpi_df.head()

LocationName  ISO      country      region world_region    MPI  \
0  Badakhshan, Afghanistan  AFG  Afghanistan  Badakhshan   South Asia  0.387   
1     Badghis, Afghanistan  AFG  Afghanistan     Badghis   South Asia  0.466   
2     Baghlan, Afghanistan  AFG  Afghanistan     Baghlan   South Asia  0.300   
3       Balkh, Afghanistan  AFG  Afghanistan       Balkh   South Asia  0.301   
4      Bamyan, Afghanistan  AFG  Afghanistan      Bamyan   South Asia  0.325   

                               geo        lat        lon  
0  (36.7347725, 70.81199529999999)  36.734772  70.811995  
1         (35.1671339, 63.7695384)  35.167134  63.769538  
2         (35.8042947, 69.2877535)  35.804295  69.287754  
3         (36.7550603, 66.8975372)  36.755060  66.897537  
4         (34.8100067, 67.8212104)  34.810007  67.821210

In [3]:
loan_theme_ids_path = os.path.join(project_root, "../../kiva_data/loan_theme_ids.csv")
loan_theme_ids = pd.read_csv(loan_theme_ids_path)

loan_theme_ids.head()

id    Loan Theme ID          Loan Theme Type  Partner ID
0  638631  a1050000000skGl                  General       151.0
1  640322  a1050000000skGl                  General       151.0
2  641006  a1050000002X1ij         Higher Education       160.0
3  641019  a1050000002X1ij         Higher Education       160.0
4  641594  a1050000002VbsW  Subsistence Agriculture       336.0

In [4]:
loan_themes_by_region_path = os.path.join(project_root, "../../kiva_data/loan_themes_by_region.csv")
loan_themes_by_region = pd.read_csv(loan_themes_by_region_path)

loan_themes_by_region.head()

Partner ID               Field Partner Name                       sector  \
0           9  KREDIT Microfinance Institution  General Financial Inclusion   
1           9  KREDIT Microfinance Institution  General Financial Inclusion   
2           9  KREDIT Microfinance Institution  General Financial Inclusion   
3           9  KREDIT Microfinance Institution  General Financial Inclusion   
4           9  KREDIT Microfinance Institution  General Financial Inclusion   

     Loan Theme ID         Loan Theme Type   country forkiva  \
0  a1050000000slfi        Higher Education  Cambodia      No   
1  a10500000068jPe  Vulnerable Populations  Cambodia      No   
2  a1050000000slfi        Higher Education  Cambodia      No   
3  a10500000068jPe  Vulnerable Populations  Cambodia      No   
4  a1050000002X1Uu              Sanitation  Cambodia      No   

                  region     geocode_old  ISO  ...  amount  \
0       Banteay Meanchey  (13.75, 103.0)  KHM  ...     450   
1    Battambang Province             NaN  KHM  ...   20275   
2    Battambang Province             NaN  KHM  ...    9150   
3  Kampong Cham Province   (12.0, 105.5)  KHM  ...  604950   
4  Kampong Cham Province   (12.0, 105.5)  KHM  ...     275   

                      LocationName                      geocode  \
0       Banteay Meanchey, Cambodia  [(13.6672596, 102.8975098)]   
1    Battambang Province, Cambodia   [(13.0286971, 102.989615)]   
2    Battambang Province, Cambodia   [(13.0286971, 102.989615)]   
3  Kampong Cham Province, Cambodia  [(12.0982918, 105.3131185)]   
4  Kampong Cham Province, Cambodia  [(12.0982918, 105.3131185)]   

                                 names                        geo        lat  \
0  Banteay Meanchey Province; Cambodia  (13.6672596, 102.8975098)  13.667260   
1        Battambang Province; Cambodia   (13.0286971, 102.989615)  13.028697   
2        Battambang Province; Cambodia   (13.0286971, 102.989615)  13.028697   
3      Kampong Cham Province; Cambodia  (12.0982918, 105.3131185)  12.098292   
4      Kampong Cham Province; Cambodia  (12.0982918, 105.3131185)  12.098292   

          lon                   mpi_region                    mpi_geo  \
0  102.897510  Banteay Mean Chey, Cambodia  (13.6672596, 102.8975098)   
1  102.989615  Banteay Mean Chey, Cambodia  (13.6672596, 102.8975098)   
2  102.989615  Banteay Mean Chey, Cambodia  (13.6672596, 102.8975098)   
3  105.313119       Kampong Cham, Cambodia  (11.9924294, 105.4645408)   
4  105.313119       Kampong Cham, Cambodia  (11.9924294, 105.4645408)   

  rural_pct  
0      90.0  
1      90.0  
2      90.0  
3      90.0  
4      90.0  

[5 rows x 21 columns]

Evaluate missing data

In [5]:
datasets = [kiva_loans, mpi_df, loan_theme_ids, loan_themes_by_region]
dataset_names = ['kiva_loans', 'mpi_df', 'loan_theme_ids', 'loan_themes_by_region']

In [6]:
def find_null_values(df, name):
    print(f"\n--- Null values in {name} ---")
    total_rows = len(df)
    null_counts = df.isnull().sum()
    null_percent = (null_counts / total_rows) * 100
    null_summary = pd.DataFrame({
        'Missing Values': null_counts,
        'Percent Missing': null_percent.round(2)
    })
    print(null_summary[null_summary['Missing Values'] > 0]
          .sort_values(by='Missing Values', ascending=False))

In [7]:
for df, name in zip(datasets, dataset_names):
    find_null_values(df, name)


--- Null values in kiva_loans ---
                  Missing Values  Percent Missing
tags                      171416            25.54
region                     56800             8.46
funded_time                48331             7.20
partner_id                 13507             2.01
use                         4232             0.63
borrower_genders            4221             0.63
disbursed_time              2396             0.36
country_code                   8             0.00

--- Null values in mpi_df ---
              Missing Values  Percent Missing
lat                     1880            67.82
lon                     1880            67.82
LocationName            1788            64.50
region                  1788            64.50
MPI                     1788            64.50
ISO                     1764            63.64
country                 1764            63.64
world_region            1764            63.64

--- Null values in loan_theme_ids ---
                 Missing Values

kiva_loans is missing 8.46% of region values which is significant considering we're estimnating regional poverty levels. To handle this missing data, I will infer missing regions using available context:

step 1: check for a strong pattern between partner_ids and region  
step 2: use that mapping to fill missing regions  
step 3: drop the remaining rows with missing region

In [8]:
region_map = kiva_loans.dropna(subset=['region']).groupby(['country', 'partner_id'])['region'].agg(lambda x: x.mode()[0]).reset_index()

def fill_region(row):
    if pd.isna(row['region']):
        match = region_map[
            (region_map['country'] == row['country']) & 
            (region_map['partner_id'] == row['partner_id'])
        ]
        if not match.empty:
            return match.iloc[0]['region']
    return row['region']

kiva_loans['region'] = kiva_loans.apply(fill_region, axis=1)
kiva_loans = kiva_loans.dropna(subset=['region'])
print(kiva_loans.isnull().sum().sort_values(ascending=False))

tags                  158517
funded_time            45494
use                     3718
borrower_genders        3712
country_code               8
id                         0
posted_time                0
repayment_interval         0
lender_count               0
term_in_months             0
disbursed_time             0
partner_id                 0
funded_amount              0
currency                   0
region                     0
country                    0
sector                     0
activity                   0
loan_amount                0
date                       0
dtype: int64


In [9]:
import pandas as pd
import numpy as np
import re
from sklearn.feature_extraction.text import CountVectorizer

kiva_loans = kiva_loans.copy()

kiva_loans['use'] = kiva_loans['use'].fillna('')

def clean_text(text):
    text = text.lower()                         # Lowercase
    text = re.sub(r'[^a-z\s]', '', text)        # Remove punctuation and numbers
    return text

kiva_loans['clean_use'] = kiva_loans['use'].apply(clean_text)

vectorizer = CountVectorizer(
    stop_words='english',       
    max_features=1000           
)

X = vectorizer.fit_transform(kiva_loans['clean_use'])

bow_df = pd.DataFrame(X.toarray(), columns=vectorizer.get_feature_names_out())

bow_df.head()


able  access  accessories  ace  acquire  activities  activity  add  adding  \
0     0       0            0    0        0           0         0    0       0   
1     0       0            0    0        0           0         0    0       0   
2     0       0            0    0        0           0         0    0       0   
3     0       0            0    0        0           0         0    0       0   
4     0       0            0    0        0           0         0    0       0   

   addition  ...  yarn  yarns  year  yeast  yield  yields  young  yucca  zinc  \
0         0  ...     0      0     0      0      0       0      0      0     0   
1         0  ...     0      0     0      0      0       0      0      0     0   
2         0  ...     0      0     0      0      0       0      0      0     0   
3         0  ...     0      0     0      0      0       0      0      0     0   
4         0  ...     0      0     0      0      0       0      0      0     0   

   zippers  
0        0  
1        0  
2        0  
3        0  
4        0  

[5 rows x 1000 columns]

Here, I preprocessed the "use" column of the "kiva_loans" dataset to prepare it for analysis. I started by filling in missing values with empty strings and cleaning the text by converting it to lowercase and removing punctuation and numbers. Next, I used CountVectorizer to transform the cleaned text into a numerical format using the Bag-of-Words model. This resulted in a new DataFrame bow_df  containing word count features for the top 1,000 most frequent words, which can be used for further analysis or modeling.

In [10]:
kiva_loans = kiva_loans[kiva_loans['use'] != ''].copy()